### PR TITLE
Add client side Kayvee error logs for golang

### PR DESCRIPTION
**Pull request by mohit**
Add a structured log in case the backend fails to respond with a successful response after the configured retries.

### CODE REVIEW

File: clients/go/gengo.go [hunk 1]
```diff
@@ -381,10 +381,15 @@ func (c *WagClient) do%sRequest(ctx context.Context, req *http.Request, headers
 	}
 	resp, err := c.requestDoer.Do(client, req)
 	if err != nil {
+		c.logger.ErrorD("client-request-finished", map[string]string{
```
Comment (kvigen) [inline clients/go/gengo.go]: do we want to put error in the title so it's clear that it's this is logged on errors.

Also, when we get a 50x, err will be nil, but resp.StatusCode will be >= 500. Do we want to track those too?

Comment (mohit) [inline clients/go/gengo.go]: I'm hoping we can in the future add this in non error conditions too, but I think we could add the `error` in the signalfx series at the matcher level. 

Yes. Thanks for pinging - it is a bug: 

TIL: ```A non-2xx status code doesn't cause an
  // error.```
https://golang.org/src/net/http/client.go?s=16536:16588#L463

Comment (mohit) [inline clients/go/gengo.go]: @kvigen here is an updated attempt: https://github.com/Clever/wag/pull/160/commits/205cd47ba9b2f625e2ff943f3f7c7a4154365db7

File: clients/js/genjs.go [hunk 2]
```diff
@@ -313,6 +322,11 @@ var methodTmplStr = `
       async.whilst(
         () => requestOptions.uri !== "",
         cbW => {
+	  this.logger.errorD("client-request-finished", {
+		"backend": "{{.ServiceName}}",
+		"message": arguments.length > 0 ? arguments[0] : "",
+		"status_code": 0
```
Comment (kvigen) [inline clients/js/genjs.go]: always status code of zero?

Comment (mohit) [inline clients/js/genjs.go]: that was just a test implementation. Actual implementation after talking with @cozmo https://github.com/Clever/wag/pull/160/commits/4c807adaa667887a308f23bae8f6b29bb017a416

File: clients/js/genjs.go [hunk 3]
```diff
@@ -338,10 +374,12 @@ var methodTmplStr = `
 
           switch (response.statusCode) {
             {{ range $response := .Responses }}case {{ $response.StatusCode }}:{{if $response.IsError }}
+			  var err = new Errors.{{ $response.Name }}(body || {});
+			  responseLog(this.logger, response, err);
```
Comment (mohit) [inline clients/js/genjs.go]: @cozmo seems like I'm having issues with `this` here. Would love to walk through with you tomorrow?

File: samples/gen-js/index.js [hunk 4]
```diff
@@ -68,6 +69,27 @@ const noRetryPolicy = {
 };
 
 /**
+ * Request status log is used to 
+ * to output the status of a request returned 
+ * by the client.
+ */
+function responseLog(logger, response, err) {
```
Comment (mohit) [inline samples/gen-js/index.js]: this is the final rendered function responseLog
 @cozmo

File: samples/gen-js/index.js [hunk 5]
```diff
@@ -188,6 +217,7 @@ class SwaggerTest {
 
       const retryPolicy = options.retryPolicy || this.retryPolicy || singleRetryPolicy;
       const backoffs = retryPolicy.backoffs();
+      const logger = this.logger;
   
       let retries = 0;
       (function requestOnce() {
```
Comment (mohit) [inline samples/gen-js/index.js]: calling responseLog @cozmo

**Comment by mohit** [comment]
Tested the node client by connecting in success and failure cases with workflow-manager.

File: clients/go/gengo.go [hunk 6]
```diff
@@ -380,11 +387,29 @@ func (c *WagClient) do%sRequest(ctx context.Context, req *http.Request, headers
 	    req = req.WithContext(ctx)
 	}
 	resp, err := c.requestDoer.Do(client, req)
+	retCode := 0
+	if resp != nil {
+	  retCode = resp.StatusCode
+	}
+
+	// log all client failures and non-successful HT
```
Comment (mohit) [inline clients/go/gengo.go]: @kvigen added `method` and `uri`

**Pull request merged**